cli/pkg/local/server.go

```diff
@@ -90,6 +103,394 @@ func (s *Server) GetVersion(ctx context.Context, r *connect.Request[localv1.GetV
+	}), nil
+}
+
+// PushToGithub assumes that the current project is not a git repo, it should generally be called after DeployValidation.
+func (s *Server) PushToGithub(ctx context.Context, r *connect.Request[localv1.PushToGithubRequest]) (*connect.Response[localv1.PushToGithubResponse], error) {
+	if !s.app.ch.IsAuthenticated() {
+		return nil, errors.New("user should be authenticated before pushing to git")
+	}
+	// Get admin client
+	c, err := client.New(s.app.adminURL, s.app.ch.AdminTokenDefault, "Rill Localhost")
+	if err != nil {
+		return nil, err
+	}
+
+	initGit := false
+	// check if project has a git repo
+	remote, _, err := gitutil.ExtractGitRemote(s.app.ProjectPath, "", false)
+	if err != nil {
+		if errors.Is(err, git.ErrRepositoryNotExists) {
+			initGit = true
+		} else if !errors.Is(err, gitutil.ErrGitRemoteNotFound) {
+			return nil, err
+		}
+	}
+	if remote != nil {
+		return nil, errors.New("git repository is already initialized with a remote")
+	}
+
+	gitStatus, err := c.GetGithubUserStatus(ctx, &adminv1.GetGithubUserStatusRequest{})
+	if err != nil {
+		return nil, err
+	}
+	if !gitStatus.HasAccess {
+		// generally this should not happen as IsGithubConnected should be true before pushing to git
+		return nil, fmt.Errorf("rill git app should be installed by user before pushing by visiting %s", gitStatus.GrantAccessUrl)
+	}
+
+	// if r.Msg.Org is empty, githubOrg will be "" which is equivalent to using default git org which is same as git username
+	var githubOrg string
+	if r.Msg.Org == "" || r.Msg.Org == gitStatus.Account {
+		githubOrg = ""
+	} else {
+		githubOrg = r.Msg.Org
+	}
+
+	// check if we have write permission on the git account
+	// this is a safety check as DeployValidation should take care of this
+	if githubOrg == "" {
+		if gitStatus.UserInstallationPermission != adminv1.GithubPermission_GITHUB_PERMISSION_WRITE {
+			return nil, fmt.Errorf("rill github app should be installed with write permission on user personal account by visiting %s", gitStatus.GrantAccessUrl)
+		}
+	} else {
+		valid := false
+		for o, p := range gitStatus.OrganizationInstallationPermissions {
+			if o == githubOrg && p == adminv1.GithubPermission_GITHUB_PERMISSION_WRITE {
+				valid = true
+				break
+			}
+		}
+		if !valid {
+			return nil, fmt.Errorf("rill github app should be installed with write permission on organization %q by visiting %s", githubOrg, gitStatus.GrantAccessUrl)
+		}
+	}
+
+	repoName := filepath.Base(s.app.ProjectPath)
+	if r.Msg.Repo != "" {
+		repoName = r.Msg.Repo
+	}
+	githubClient := github.NewTokenClient(ctx, gitStatus.AccessToken)
+	defaultBranch := "main"
+
+	// create remote repo
+	githubRepo, _, err := githubClient.Repositories.Create(ctx, githubOrg, &github.Repository{Name: &repoName, DefaultBranch: &defaultBranch})
+	if err != nil {
+		if !strings.Contains(err.Error(), "name already exists") {
+			// TODO should we retry with new name, append a number to repoName like -1, -2 etc
```

Comment:
Yes, I think that was actually also requested by Nishant

Reply:
handled, retry 3 times